Add better constructor support for deserialization

tcortega opened the pull request:
The behaviour in the mechanism is based on the behaviour observed on the System.Text.Json library, when attempting to deserialize JSON strings.

- Introduced a mechanism for dynamically identifying and invoking the most suitable constructor for a given type, making it able to call any constructor with `N` parameters. This specially useful when dealing with records. Because most people declare records as following: `record MyRecord(string MyString, ...)`
- If there are matching keys between the TOML string and the constructor's parameters, it initializes the properties from the parameters.
- Introduced a `TomlSerializerOptions` class with an `OverrideConstructorValues` option.

if there is anything wrong with it, let me know or feel free to change it yourself.

SamboyCoding commented on Tomlet/Extensions/ReflectionExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Tomlet.Extensions
+{
+    public static class ReflectionExtensions
```
Please make extension classes internal so they don't get exposed to consumers of the library

SamboyCoding replied:
Please make extension classes internal so they don't get exposed to consumers of the library

SamboyCoding commented on Tomlet/Extensions/StringExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Tomlet.Extensions
+{
+    public static class StringExtensions
```
Likewise, extension class, internal please.

SamboyCoding commented on Tomlet/GenericExtensions.cs:
```diff
@@ -8,7 +8,7 @@
 
 namespace Tomlet
```
Should this not have been moved to the Tomlet.Extensions namespace and folder?

SamboyCoding commented on Tomlet/TomlCompositeDeserializer.cs:
```diff
@@ -37,53 +38,41 @@ public static TomlSerializationMethods.Deserialize<object> For(Type type)
 
             //Ignore NonSerialized and CompilerGenerated fields.
-            fields = fields.Where(f => !f.IsNotSerialized && f.GetCustomAttribute<CompilerGeneratedAttribute>() == null).ToArray();
+            fields = fields.Where(f => !f.IsNotSerialized && f.GetCustomAttribute<CompilerGeneratedAttribute>() == null)
```
Please don't make formatting changes

SamboyCoding commented on Tomlet/TomlCompositeDeserializer.cs:
```diff
@@ -37,53 +38,41 @@ public static TomlSerializationMethods.Deserialize<object> For(Type type)
                 .Where(p => p.GetSetMethod(true) != null && p.GetCustomAttribute<TomlNonSerializedAttribute>() == null)
-                .Select(p => new KeyValuePair<PropertyInfo, TomlPropertyAttribute?>(p, p.GetCustomAttribute<TomlPropertyAttribute>()))
+                .Select(p =>
+                    new KeyValuePair<PropertyInfo, TomlPropertyAttribute?>(p,
```
Formatting change

SamboyCoding commented on Tomlet/TomlCompositeDeserializer.cs:
```diff
@@ -121,4 +114,47 @@ public static TomlSerializationMethods.Deserialize<object> For(Type type)
+        if (parameters.Length == 0)
+        {
+            assignedMembers = new HashSet<string>();
+            return Activator.CreateInstance(type)!;
```
We're no longer catching any exceptions that occur in CreateInstance and wrapping them in a TomlInstantiationException.

tcortega replied:
There aren't really any cases in which CreateInstance would fail considering we're already filtering the classes' constructor on the `TryGetBestMatchConstructor`.

I am actually replacing it to a simple `constructor.Invoke` since it'll be a cheaper call.

tcortega replied:
Also, if there is exception throwing within the constructor, it's not really the Serializer's fault or concern.

SamboyCoding commented on Tomlet/TomlCompositeDeserializer.cs:
```diff
@@ -121,4 +114,47 @@ public static TomlSerializationMethods.Deserialize<object> For(Type type)
+            return Activator.CreateInstance(type)!;
+        }
+
+        if (tomlValue is not TomlTable table)
```
I would do this before TryGetBestMatchConstructor, as that method seems moderately expensive, so get the cheap checks out of the way first.

SamboyCoding commented on Tomlet/TomlCompositeDeserializer.cs:
```diff
@@ -121,4 +114,47 @@ public static TomlSerializationMethods.Deserialize<object> For(Type type)
+            var parameter = parameters[i];
+            object argument;
+            
+            if (!table.TryGetValue(parameter.Name.ToPascalCase(), out var entry))
```
Compiler warning on this line, please fix

SamboyCoding commented on Tomlet/TomlCompositeSerializer.cs:
```diff
@@ -33,8 +33,8 @@ public static TomlSerializationMethods.Serialize<object> For(Type type)
             fields = fields.Where(f => !(f.IsNotSerialized || f.GetCustomAttribute<TomlNonSerializedAttribute>() != null)
-                && f.GetCustomAttribute<CompilerGeneratedAttribute>() == null 
-                && !f.Name.Contains('<')).ToArray();
+                                       && f.GetCustomAttribute<CompilerGeneratedAttribute>() == null 
```
Formatting change, please revert

SamboyCoding commented on Tomlet/TomlSerializationMethods.cs:
```diff
@@ -75,8 +75,10 @@ static TomlSerializationMethods()
-        internal static Serialize<object> GetSerializer(Type t)
+        internal static Serialize<object> GetSerializer(Type t, TomlSerializerOptions? options)
         {
+            options ??= new TomlSerializerOptions();
```
Given that this is a fairly often-called method, can you make a static TomlSerializerOptions.Default and use that instead here, to avoid an allocation.

SamboyCoding commented on Tomlet/TomlSerializationMethods.cs:
```diff
@@ -104,39 +106,41 @@ internal static Serialize<object> GetSerializer(Type t)
-        internal static Deserialize<object> GetDeserializer(Type t)
+        internal static Deserialize<object> GetDeserializer(Type t, TomlSerializerOptions? options)
         {
+            options ??= new TomlSerializerOptions();
```
See above w.r.t. allocations

SamboyCoding commented on Tomlet/TomlSerializationMethods.cs:
```diff
@@ -146,14 +150,14 @@ internal static Deserialize<object> GetDeserializer(Type t)
                 {
 #if NETFRAMEWORK
-                    return (Deserialize<object>)_stringKeyedDictionaryMethod.MakeGenericMethod(genericArgs[1]).Invoke(null, new object[0])!;
+                    return (Deserialize<object>)_stringKeyedDictionaryMethod.MakeGenericMethod(genericArgs[1]).Invoke(null, new object[]{options})!;
```
As this is no longer conditionally using Array.Empty, the #if can be removed

SamboyCoding commented on Tomlet/TomletMain.cs:
```diff
@@ -12,63 +11,64 @@ namespace Tomlet
     {
         [Attributes.ExcludeFromCodeCoverage]
-        public static void RegisterMapper<T>(TomlSerializationMethods.Serialize<T>? serializer, TomlSerializationMethods.Deserialize<T>? deserializer)
+        public static void RegisterMapper<T>(TomlSerializationMethods.Serialize<T>? serializer,
```
Formatting change, please revert

tcortega committed:
Add better constructor support for deserialization

The behaviour in the mechanism is based on the behaviour observed on the System.Text.Json library, when attempting to deserialize JSON strings.

Introduced a mechanism for dynamically identifying and invoking the most suitable constructor for a given type, making it able to call any constructor with N parameters. This specially useful when dealing with records. Because most people declare records as following: record MyRecord(string MyString, ...)
If there are matching keys between the TOML string and the constructor's parameters, it initializes the properties from the parameters.
Introduced a TomlSerializerOptions class with an OverrideConstructorValues option.

tcortega commented:
When I moved the GenericExtensions as you said, to the Extensions folder, there seemed to be an issue with using the Extension methods in the `TomlCompositeDeserializer.cs`, an error saying.
```
Ambiguous invocation:   Tomlet.Attributes.TomlNonSerializedAttribute GetCustomAttribute<TomlNonSerializedAttribute>(this System.Reflection.MemberInfo) (in class CustomAttributeExtensions)   Tomlet.Attributes.TomlNonSerializedAttribute? GetCustomAttribute<TomlNonSerializedAttribute>(this System.Reflection.MemberInfo) (in class GenericExtensions) match
```
Therefore it is now calling the method in the static class directly.
